src/vaults/SectorCrossVault.sol

```diff
@@ -0,0 +1,284 @@
+		internal
+		pure
+		returns (UserRequest memory userRequest)
+	{
+		bytes memory callDataWithoutSelector = slice(_data, 4, _data.length - 4);
+		(userRequest) = abi.decode(callDataWithoutSelector, (UserRequest));
+	}
+
+	/// @notice Decodes and verifies socket calldata
+	/// @param _data Bungee txn calldata
+	/// @param _chainId chainId to check in bungee calldata
+	/// @param _inputToken inputWantToken to check in bungee calldata
+	/// @param _receiverAddress receiving address to check in bungee calldata
+	function verifySocketCalldata(
+		bytes memory _data,
+		uint256 _chainId,
+		address _inputToken,
+		address _receiverAddress
+	) internal view {
+		UserRequest memory userRequest;
+		(userRequest) = decodeSocketRegistryCalldata(_data);
+
+		if (userRequest.toChainId != _chainId) {
+			revert("Invalid chainId");
+		}
+		if (userRequest.receiverAddress != _receiverAddress) {
+			revert("Invalid receiver address");
+		}
+		if (userRequest.bridgeRequest.inputToken != _inputToken) {
+			revert("Invalid input token");
+		}
+
+		if (!sectorVaultsWhitelist[_chainId][userRequest.receiverAddress]) revert ReceiverNotWhiteslisted(_receiverAddress);
+	}
+
+	// Who should be responsible for whitelist vaults?
+	// I believe it's the guardian
+	function whitelistSectorVault(uint32 chainId, address _vault) external onlyRole(GUARDIAN) {
+		sectorVaultsWhitelist[chainId][_vault] = true;
+		emit WhitelistedSectorVault(chainId, _vault);
+	}
+
+	function checkWhitelistVault(uint32 chainId, address vault) external view returns (bool) {
+		return sectorVaultsWhitelist[chainId][vault];
+	}
+
+	// Added function to emit event
+	// This one has to integrate with layerZero message sender
+	function startDepositCrosschainRequest(
+		uint32 _fromChainId,
+		uint32 _toChainId,
+		uint256 amount
+	) public {
+		emit BridgeAsset(_fromChainId, _toChainId, amount);
+	}
+
+	// This function will change to depositCrossChain
+    /// @notice Sends tokens using Bungee middleware. Assumes tokens already present in contract. Manages allowance and transfer.
+    /// @dev Currently not verifying the middleware request calldata. Use very carefully
+    /// @param allowanceTarget address to allow tokens to swipe
+    /// @param socketRegistry address to send bridge txn to
+    /// @param destinationAddress address of receiver
+    /// @param amount amount of tokens to bridge
+    /// @param destinationChainId chain Id of receiving chain
+    /// @param data calldata of txn to be sent
+    function sendTokens(
+        address allowanceTarget,
+        address socketRegistry,
+        address destinationAddress,
+        uint256 amount,
+        uint256 destinationChainId,
+        bytes calldata data
+    ) public onlyRole(MANAGER) {
+        verifySocketCalldata(
+            data,
+            destinationChainId,
+            address(_asset),
+            destinationAddress
+        );
+
+		_asset.approve(msg.sender, amount);
```

Comment:
why do we need this approval?

Reply:
I don't get it too, but the transaction was always falling without it. In the brahma finance example they don't do it but in their case the user calling sendTokens is the owner of the assets. Our case the vault is the owner. My logic was since the msg.sender here is the manager and it doesn't own the tokens it has to approve it before doing anything else.

Reply:
Makes sense.
Is it because data contains a signed tx by manger and that's what socketRegistry registry is using?